consensus/cometbft/service/abci.go

```diff
@@ -36,11 +36,29 @@ var (
 	errNilFinalizeBlockState = errors.New("finalizeBlockState is nil")
 )
 
+// combineContexts combines the CometBFT context with the original context from the node.
+// TODO: This can be removed if CometBFT implements contexts.
+func (s *Service) combineContexts(ctx context.Context) context.Context {
+	if s.ctx == nil {
+		return ctx
+	}
+	newCtx, cancel := context.WithCancel(context.Background())
+	go func() {
+		select {
+		case <-ctx.Done():
+			cancel()
+		case <-s.ctx.Done():
+			cancel()
+		}
```

Comment:
@shotes: is this leaking goroutines? 
We create a goroutine per each callback and never exit them.
Probably we need to make things slighly more complex.
We want to:
- have a single goroutine monitoring contextes
- reset the context from consensus every time we have a callback

Reply:
An possible alternative is to return a channel allowing to close the goroutine?

Something like:

```
	done := make(chan struct{})
	newCtx, cancel := context.WithCancel(context.Background())
	go func() {
		select {
		case <-ctxConsensus.Done():
			cancel()
		case <-ctxApp.Done():
			cancel()
		case <-done:
			return
		}
	}()
	return newCtx, done
```

and then change callback as:

```
	ctx, done := combineContexts(ctxConsensus, s.ctx)
	defer func() { close(done) }()
	return s.finalizeBlock(ctx, req)
```

I am not 100% sure this is good, cause I do not know what happens if `s.Ctx` is cancelled among callbacks, when no goroutine is spinning.

Reply:
@rezbera @shotes I tried this and I still got a crash.
May it be that given that in simulated tests, node.Start is called in a different go-routine there is race?
We sleep two seconds after start, maybe not enough?